src/js/view/controls/components/menu/settings-menu.js

```diff
@@ -3,262 +3,463 @@ import { MenuItem } from 'view/controls/components/menu/menu-item';
+        model.on('change:playlistItem', this.onPlaylistItem, this);
+        model.change('captionsIndex', this.onCaptionsIndex, this);
+        model.change('playbackRates', this.onPlaybackRates, this); 
+        model.change('playbackRate', this.onPlaybackRate, this); 
+        model.on('change:playbackRateControls', this.onPlaybackRateControls, this);
+        model.on('change:castActive', this.onCastActive, this);
+        model.on('change:streamType', this.onChangeStreamType, this);
+    }
+
+    open(evt) {
+        if (this.visible) {
+            return;
+        }
+        this.el.parentNode.classList.add('jw-settings-open');
+        this.trigger('menuVisibility', { visible: true, evt });
+        document.addEventListener('click', this.onDocumentClick);
+        this.el.setAttribute('aria-expanded', 'true');
+        this.visible = true;
+    }
+
+    close(evt) {
+        this.el.parentNode.classList.remove('jw-settings-open');
+        this.trigger('menuVisibility', { visible: false, evt });
+        document.removeEventListener('click', this.onDocumentClick);
+        this.visible = false;
+        if (this.openMenus.length) {
+            this.closeChildren();
+        }
+
+        // If closed by keypress, focus appropriate element.
+        const key = normalizeKey(evt && evt.sourceEvent && evt.sourceEvent.key);
+        const gearButton = this.controlbar.elements.settingsButton.element();
+        let focusEl;
+
+        switch (key) {
+            case 'Right':
+                focusEl = nextSibling(gearButton);
+                break;
+            case 'Left':
+                focusEl = previousSibling(gearButton);
+                break;
+            case 'Tab':
+                if (evt.shiftKey) {
+                    focusEl = previousSibling(gearButton);
+                    break;
+                }
+                focusEl = gearButton;
+                break;
+            default:
+                break;
+        }
+        if (focusEl) {
+            focusEl.focus();
+        }
+    }
+
+    updateControlbarButtons(menuName) {
+        const { children, controlbar, model } = this;
+        const shouldShowGear = 
+            !!children.quality || 
+            !!children.playbackRates || 
+            Object.keys(children).length > 1;
+
+        controlbar.elements.settingsButton.toggle(shouldShowGear);
+        if (children.captions) {
+            controlbar.toggleCaptionsButtonState(!!model.get('captionsIndex'));
+        }
+        const controlBarButton = controlbar.elements[`${menuName}Button`];
+        if (controlBarButton) {
+            const isVisible = !!children[menuName];
+            controlBarButton.toggle(isVisible);
+        }
+    }
+
+    destroy() {
+        Object.keys(this.children).map(menuName => {
+            this.children[menuName].destroy();
+        });
+        document.removeEventListener('click', this.onDocumentClick);
+        if (this.model) {
```

Comment:
`this.model` will always be defined if passed in to the constructor. Set `this.model` to `null` here after calling `this.model.off`.